samples/nrf9160/modem_fault/src/main.c

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <string.h>
+#include <stdlib.h>
+#include <zephyr/kernel.h>
+#include <net/socket.h>
+#include <net/socket.h>
+#include <modem/nrf_modem_lib.h>
+#include <modem/lte_lc.h>
+#include <nrf_modem.h>
+#include <nrf_socket.h>
+
+LOG_MODULE_REGISTER(app);
+
+#define HTTP_HOSTNAME "speedtest.ftp.otenet.gr"
+#define HTTP_PORT "80"
+#define HTTP_GET                                                                                   \
+	"GET /files/test100k.db HTTP/1.1\r\n"                                                      \
+	"Host: speedtest.ftp.otenet.gr\r\n"                                                        \
+	"Connection: close\r\n"                                                                    \
+	"\r\n"
+
+#define HTTP_GET_LEN (sizeof(HTTP_GET) - 1)
+#define HTTP_HDR_END "\r\n\r\n"
+
+static const char send_buf[] = HTTP_GET;
+static char recv_buf[2048];
+
+/* Modem library initialization semaphore */
+static K_SEM_DEFINE(modem_sem, 0, 1);
+/* Network registration semaphore, given by main thread */
+static K_SEM_DEFINE(cereg_sem, 0, 1);
+
+/* Workqueue task crashing the modem by shutting it down
+ * while a download is in progress, without first de-registring from the network.
+ *
+ * In "normal" crash scenarions, the modem library remains initialized
+ * until it is shutdown by the default fault handler or by the application.
+ * As long as it remains initialized, it is possible to read out any
+ * pending network data that was delivered by modem before it crashed.
+ */
+static void crash_task(struct k_work *unused)
+{
+	(void) nrf_modem_shutdown();
+}
+
+static K_WORK_DELAYABLE_DEFINE(crash_work, crash_task);
+
+/* Modem library initialization hook */
+NRF_MODEM_LIB_ON_INIT(on_init, on_modem_init, NULL);
+
+static void on_modem_init(int err, void *ctx)
+{
+	if (err) {
+		LOG_INF("on_init: initialization failed, err %d", err);
+		return;
+	}
+
+	LOG_INF("on_init: modem library initialized");
+
+	/* Let main thread continue and register the device
+	 * to the LTE network upon modem re-initialization.
+	 */
+	k_sem_give(&modem_sem);
+}
+
+static void network_thread(void *p1, void *p2, void *p3)
+{
+	int err;
+	int fd;
+	size_t off;
+	ssize_t bytes;
+	struct addrinfo *res;
+	struct addrinfo hints = {
+		.ai_family = AF_INET,
+		.ai_socktype = SOCK_STREAM,
+		.ai_protocol = IPPROTO_TCP,
+	};
+
+	bool has_faulted = false;
+
+wait_for_cereg:
+	LOG_INF("net: waiting for network registration");
+	k_sem_take(&cereg_sem, K_FOREVER);
+
+	err = getaddrinfo(HTTP_HOSTNAME, HTTP_PORT, &hints, &res);
+	if (err) {
+		LOG_ERR("net: getaddrinfo() failed, err %d", errno);
+		return;
+	}
+
+	fd = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
+	if (fd == -1) {
+		LOG_ERR("net: failed to open socket!");
```

Comment:
Does not make much difference in this sample, but `freeaddrinfo()` should be called in case of failure.